### PR TITLE
Configurable log format

### PR DESCRIPTION
**Pull request by bzon**
# Propose changes

Make structured logging configurable in JSON format.

### CODE REVIEW

File: cmd/fluxd/main.go [hunk 1]
```diff
@@ -192,7 +193,15 @@ func main() {
 	// Logger component.
 	var logger log.Logger
 	{
-		logger = log.NewLogfmtLogger(os.Stderr)
+		switch *logFormat {
+		case "json":
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stdout))
```
Comment (stefanprodan) [inline cmd/fluxd/main.go]: Please keep the same output as before `Stderr`

File: cmd/fluxd/main.go [hunk 2]
```diff
@@ -192,7 +193,15 @@ func main() {
...
+		case "fmt":
+			logger = log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+		default:
+			logger.Log("err", fmt.Sprintf("log-format %s is not valid", *logFormat))
```
Comment (stefanprodan) [inline cmd/fluxd/main.go]: The default should be `fmt` for backwards compatibly.

File: cmd/fluxd/main.go [hunk 3]
```diff
@@ -98,6 +98,7 @@ func main() {
 	}
 	// This mirrors how kubectl extracts information from the environment.
 	var (
+		logFormat         = fs.String("log-format", "default|json", "change the log format.")
```
Comment (stefanprodan) [inline cmd/fluxd/main.go]: The default should be fmt for backwards compatibly.

File: cmd/helm-operator/main.go [hunk 4]
```diff
@@ -78,6 +80,8 @@ func init() {
 
 	versionFlag = fs.Bool("version", false, "print version and exit")
 
+	logFormat = fs.String("log-format", "default|json", "change the log format.")
```
Comment (stefanprodan) [inline cmd/helm-operator/main.go]: The default should be fmt for backwards compatibly.

File: cmd/helm-operator/main.go [hunk 5]
```diff
@@ -116,7 +120,15 @@ func main() {
...
 	// init go-kit log
 	{
-		logger = log.NewLogfmtLogger(os.Stderr)
+		switch *logFormat {
+		case "json":
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stdout))
+		case "fmt":
+			logger = log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
+		default:
+			logger.Log("err", fmt.Sprintf("log-format %s is not valid", *logFormat))
```
Comment (stefanprodan) [inline cmd/helm-operator/main.go]: The default should be fmt for backwards compatibly.

File: cmd/helm-operator/main.go [hunk 6]
```diff
@@ -116,7 +120,15 @@ func main() {
 
 	// init go-kit log
 	{
-		logger = log.NewLogfmtLogger(os.Stderr)
+		switch *logFormat {
+		case "json":
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stdout))
```
Comment (stefanprodan) [inline cmd/helm-operator/main.go]: Please keep the same output as before Stderr

File: chart/flux/values.yaml [hunk 7]
```diff
@@ -5,6 +5,8 @@ token: ""
 
 replicaCount: 1
 
+logFormat: default
```
Comment (stefanprodan) [inline chart/flux/values.yaml]: This should be fmt

File: chart/flux/templates/helm-operator-deployment.yaml [hunk 8]
```diff
@@ -118,6 +118,7 @@ spec:
           mountPath: /var/fluxd/helm/repository/cache
         {{- end }}
         args:
+        - --log-format={{ .Values.logFormat }}
```
Comment (stefanprodan) [inline chart/flux/templates/helm-operator-deployment.yaml]: This should be set only if `.Values.logFormat` exists otherwise this introduces a breaking change.

File: chart/flux/templates/deployment.yaml [hunk 9]
```diff
@@ -148,6 +148,7 @@ spec:
 {{ toYaml .Values.extraEnvs | indent 10 }}
           {{- end }}
           args:
+          - --log-format={{ .Values.logFormat }}
```
Comment (stefanprodan) [inline chart/flux/templates/deployment.yaml]: This should be set only if `.Values.logFormat` exists otherwise this introduces a breaking change.

File: chart/flux/README.md [hunk 10]
```diff
@@ -188,6 +188,7 @@ The following tables lists the configurable parameters of the Weave Flux chart a
 | `replicaCount`                                    | `1`                                                  | Number of Flux pods to deploy, more than one is not desirable.
 | `image.pullPolicy`                                | `IfNotPresent`                                       | Image pull policy
 | `image.pullSecret`                                | `None`                                               | Image pull secret
+| `logFormat`                                       | `default`                                            | Log format (default or json)
```
Comment (stefanprodan) [inline chart/flux/README.md]: The default should be `fmt`

File: chart/flux/Chart.yaml [hunk 11]
```diff
@@ -1,6 +1,6 @@
 apiVersion: v1
 appVersion: "1.12.3"
-version: 0.9.5
+version: 0.9.6
```
Comment (stefanprodan) [inline chart/flux/Chart.yaml]: Please undo this change, the chart version bump will be done in the release PR.

**Comment by bzon** [comment]
@stefanprodan the requested changes have been applied accordingly.

File: cmd/fluxd/main.go [hunk 12]
```diff
@@ -192,7 +193,14 @@ func main() {
 	// Logger component.
 	var logger log.Logger
 	{
-		logger = log.NewLogfmtLogger(os.Stderr)
+		switch *logFormat {
+		case "json":
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
+		case "fmt":
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
```
Comment (squaremo) [inline cmd/fluxd/main.go]: Should this be `NewLogfmtLogger`?

File: cmd/helm-operator/main.go [hunk 13]
```diff
@@ -116,7 +120,14 @@ func main() {
 
 	// init go-kit log
 	{
-		logger = log.NewLogfmtLogger(os.Stderr)
+		switch *logFormat {
+		case "json":
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
+		case "fmt":
+			logger = log.NewJSONLogger(log.NewSyncWriter(os.Stderr))
```
Comment (hiddeco) [inline cmd/helm-operator/main.go]: I think this should be

```suggestion
			logger = log.NewLogfmtLogger(log.NewSyncWriter(os.Stderr))
```

File: cmd/fluxd/main.go (hunk 12 above)
Comment (hiddeco) [inline cmd/fluxd/main.go]: Same here.

Comment (bzon) [reply cmd/fluxd/main.go]: ahhh right. 😓

Comment (bzon) [reply cmd/fluxd/main.go]: I amended the commit to fix this. 👍
